File: demos/pax-britannica/pax-britannica/src/de/swagner/paxbritannica/factory/Production.java
package de.swagner.paxbritannica.factory;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;

import de.swagner.paxbritannica.Constants;
import de.swagner.paxbritannica.GameInstance;
import de.swagner.paxbritannica.Resources;
import de.swagner.paxbritannica.bomber.Bomber;
import de.swagner.paxbritannica.fighter.Fighter;
import de.swagner.paxbritannica.frigate.Frigate;

public class Production {

	public float BUILDING_SPEED = 100;
	
	public int MAXSHIPS = 100;

	private float SEGMENTS = 32;
	private float RADIUS = 32;
	private float DRAW_OFFSET = -4;
	private float SPAWN_OFFSET = 47;

	public float fighterCost = 50;
	public float bomberCost = 170;
	public float frigateCost = 360;
	public float upgradeCost = 1080;

	private float potential_cost = 0;
	private float needle_angle = 0;
	private float needle_velocity = 0;

	private Vector2 facing90 = new Vector2();

	public boolean halt_production = false;

	private Sprite needle;

	private Sprite upgrade_outline;
	private Sprite frigate_outline;
	private Sprite bomber_outline;
	private Sprite fighter_outline;

	private Sprite production1;
	private Sprite production2;
	private Sprite production3;
	
	private Sprite production_tile1;
	private Sprite production_tile2;
	private Sprite production_tile3;
	private Sprite production_tile4;

	private Sprite health_none;
	private Sprite health_some;
	private Sprite health_full;
	
	public int currentBuildingUnit = -1;
	
	public float fade = 0;
	
//	public ImmediateModeRenderer renderer = new ImmediateModeRenderer();
//	OrthographicCamera camera;

	private FactoryProduction factory;

	public Production(FactoryProduction factory) {
		this.factory = factory;
		
//		camera = new OrthographicCamera(800, 480);
//		camera.translate(400, 240, 0);
		
		needle = Resources.getInstance().needle;
		needle.setOrigin(needle.getHeight() / 2, needle.getWidth() / 2);

		production1 = Resources.getInstance().production1;
		production1.setOrigin(production1.getHeight() / 2, production1.getWidth() / 2);
		production2 = Resources.getInstance().production2;
		production2.setOrigin(production2.getHeight() / 2, production2.getWidth() / 2);
		production3 = Resources.getInstance().production3;
		production3.setOrigin(production3.getHeight() / 2, production3.getWidth() / 2);
		
		production_tile1 = Resources.getInstance().production_tile1;
		production_tile1.setOrigin(production_tile1.getHeight() / 2, production_tile1.getWidth() / 2);
		production_tile2 = Resources.getInstance().production_tile2;
		production_tile2.setOrigin(production_tile2.getHeight() / 2, production_tile2.getWidth() / 2);
		production_tile3 = Resources.getInstance().production_tile3;
		production_tile3.setOrigin(production_tile3.getHeight() / 2, production_tile3.getWidth() / 2);
		production_tile4 = Resources.getInstance().production_tile4;
		production_tile4.setOrigin(production_tile4.getHeight() / 2, production_tile4.getWidth() / 2);

		upgrade_outline = Resources.getInstance().upgradeOutline;
		upgrade_outline.setOrigin(upgrade_outline.getHeight() / 2, upgrade_outline.getWidth() / 2);
		frigate_outline = Resources.getInstance().frigateOutline;
		frigate_outline.setOrigin(frigate_outline.getHeight() / 2, frigate_outline.getWidth() / 2);
		bomber_outline = Resources.getInstance().bomberOutline;
		bomber_outline.setOrigin(bomber_outline.getHeight() / 2, bomber_outline.getWidth() / 2);
		fighter_outline = Resources.getInstance().fighterOutline;
		fighter_outline.setOrigin(fighter_outline.getHeight() / 2, fighter_outline.getWidth() / 2);

		health_none = Resources.getInstance().healthNone;
		health_none.setOrigin(health_none.getHeight() / 2, health_none.getWidth() / 2);
		health_some = Resources.getInstance().healthSome;
		health_some.setOrigin(health_some.getHeight() / 2, health_some.getWidth() / 2);
		health_full = Resources.getInstance().healthFull;
		health_full.setOrigin(health_full.getHeight() / 2, health_full.getWidth() / 2);
	}

	/**
	 * Span new Ship
	 * 
	 * 1 = Fighter
	 * 2 = Bomber
	 * 3 = Frigate
	 * 4 = Upgrade
	 * @param unitType
	 */
	public void spawn(int unitType) {
//		factory.ownShips = 0;
//		for (Ship fighter : GameInstance.getInstance().fighters) {
//			if(fighter.id == factory.id) factory.ownShips++;
//		}		
//		for (Ship bomber : GameInstance.getInstance().bombers) {
//			if(bomber.id != factory.id) factory.ownShips++;
//		}		
//		for (Ship frigate : GameInstance.getInstance().frigates) {
//			if(frigate.id != factory.id) factory.ownShips++;
//		}
//		if(factory.ownShips>MAXSHIPS) {
//			return;
//		}
		
		
		Vector2 spawn_pos = new Vector2(factory.collisionCenter.x + (SPAWN_OFFSET * factory.facing.x), 
				factory.collisionCenter.y + (SPAWN_OFFSET * factory.facing.y));

		if (unitType == 1) {
			factory.resourceAmount -= fighterCost;
			GameInstance.getInstance().fighters.add(new Fighter(factory.id, spawn_pos, new Vector2(factory.facing.x, factory.facing.y)));
		} else if (unitType == 2) {
			spawn_pos.sub(10, 10);
			factory.resourceAmount -= bomberCost;
			GameInstance.getInstance().bombers.add(new Bomber(factory.id, spawn_pos, new Vector2(factory.facing.x, factory.facing.y)));
		} else if (unitType == 3) {
			spawn_pos.sub(25, 25);
			factory.resourceAmount -= frigateCost;
			GameInstance.getInstance().frigates.add(new Frigate(factory.id, spawn_pos, new Vector2(factory.facing.x, factory.facing.y)));
		} else if (unitType == 4) {
			factory.resourceAmount -= upgradeCost;
			factory.upgradesUsed += 1;
			factory.harvestRate += (factory.harvestRateUpgrade/factory.upgradesUsed);
		}
	}

	private void update() {
		if (factory.button_held) {
			if (potential_cost == 0 && factory.resourceAmount >= potential_cost) {
				potential_cost = fighterCost;
			}
			if (factory.resourceAmount > potential_cost + (BUILDING_SPEED * Gdx.graphics.getDeltaTime()) - 1) {
				potential_cost += BUILDING_SPEED * Gdx.graphics.getDeltaTime();
			}
		} else {
			if (potential_cost > upgradeCost) {
				spawn(4);
			} else if (potential_cost > frigateCost) {
				spawn(3);
			} else if (potential_cost > bomberCost) {
				spawn(2);
			} else if (potential_cost > fighterCost) {
				spawn(1);
			}
			potential_cost = 0;
		}
	}

	public float scale_angle(float frames) {
		float angle = 0;
		if (frames < fighterCost) {
			angle = 0.25f;
		} else if (frames < bomberCost) {
			angle = (frames - fighterCost) / (bomberCost - fighterCost) * 0.25f + 0.25f;
		} else if (frames < frigateCost) {
			angle = (frames - bomberCost) / (frigateCost - bomberCost) * 0.25f + 0.5f;
		} else {
			angle = (frames - frigateCost) / (upgradeCost - frigateCost) * 0.25f + 0.75f;
		}
		return Math.min(angle - 0.25f, 1);
	}

	public float get_resources_spent(float frames) {
		float spent = 0;
		if (frames < fighterCost) {
			spent = 0;
		} else if (frames < bomberCost) {
			spent = fighterCost;
		} else if (frames < frigateCost) {
			spent = bomberCost;
		} else if (frames < upgradeCost) {
			spent = frigateCost;
		} else {
			spent = upgradeCost;
		}
		return spent;
	}

	public void draw(SpriteBatch spriteBatch) {
		update();

		facing90.set(factory.facing);
		facing90.rotate(90).nor();
//		factory.facing.nor();

//		//Renders behind sprites... :(
//		float SEGMENTS = 32;
//		float RADIUS = 32;
//		renderer.begin(GL10.GL_TRIANGLE_FAN);
//		Gdx.graphics.getGL11().glPushMatrix();
//		camera.position.set(350, +240, 0);
//		camera.update();
//		camera.apply(Gdx.graphics.getGL10());
//		Gdx.graphics.getGL11().glTranslatef(factory.collisionCenter.x,
//				factory.collisionCenter.y, 0);
//		Gdx.graphics.getGL11().glRotatef(factory.facing.angle() - 90, 0, 0, 1);
//		float angle = factory.production
//				.scale_angle(factory.resourceAmount) * MathUtils.PI * 2f;
//		float filled_angle = MathUtils.PI * 2f - angle;
//		renderer.color(0, 0, 0, 1.6f);
//		renderer.vertex(0, 0, -10);
//		for (int point = 0; point < SEGMENTS; point++) {
//			Vector2 vert = new Vector2(MathUtils.cos(MathUtils.PI / 2.f - angle - point / SEGMENTS * filled_angle) * RADIUS, MathUtils.sin(MathUtils.PI / 2.f
//					- angle - point / SEGMENTS * filled_angle)
//					* RADIUS);
//			renderer.vertex(vert.x, vert.y, -10);
//		}
//		renderer.vertex(0, 0, 10);
//		renderer.end();
//
//		Gdx.graphics.getGL11().glPopMatrix();
//		camera.position.set(400, 240, 0);
//		camera.update();
//		camera.apply(Gdx.graphics.getGL10());
		
		float angle = Math.min(1, Math.max(0, norm(scale_angle(factory.resourceAmount),0,0.25f)));
		fade = Math.min(MathUtils.PI, fade+(Gdx.graphics.getDeltaTime()));
		if(fade == MathUtils.PI) fade =0;
		
		
		production_tile1.setOrigin(0, 0);
		
//		if(factory.ownShips>MAXSHIPS-1) {
//			production_tile1.setColor(1.0f, 0.1f,0.1f, MathUtils.sin(fade));
//		} else {
			production_tile1.setColor(angle, angle,angle, angle);
//		}
		production_tile1.setRotation(factory.facing.angle());
		production_tile1.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32* facing90.x), 
				factory.collisionCenter.y	- (35 * factory.facing.y) - (32 * facing90.y));
		production_tile1.draw(spriteBatch);
		
		angle = Math.min(1, Math.max(0, norm(scale_angle(factory.resourceAmount),0.25f,0.5f)));
		production_tile2.setOrigin(0, 0);
//		if(factory.ownShips>MAXSHIPS-1) {
//			production_tile2.setColor(1, 0.1f,0.1f, MathUtils.sin(fade));
//		} else {
			production_tile2.setColor(angle, angle,angle, angle);
//		}
		production_tile2.setRotation(factory.facing.angle());
		production_tile2.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
				factory.collisionCenter.y	- (35 * factory.facing.y) - (32 * facing90.y));
		production_tile2.draw(spriteBatch);
		
		angle = Math.min(1, Math.max(0, norm(scale_angle(factory.resourceAmount),0.5f,0.75f)));
		production_tile3.setOrigin(0, 0);
//		if(factory.ownShips>MAXSHIPS-1) {
//			production_tile3.setColor(1, 0.1f,0.1f, MathUtils.sin(fade));
//		} else {
			production_tile3.setColor(angle, angle,angle, angle);
//		}
		production_tile3.setRotation(factory.facing.angle());
		production_tile3.setPosition(factory.collisionCenter.x - (35* factory.facing.x)- (32 * facing90.x), 
				factory.collisionCenter.y	- (35 * factory.facing.y) - (32 * facing90.y));
		production_tile3.draw(spriteBatch);
		
		angle = Math.min(1, Math.max(0, norm(scale_angle(factory.resourceAmount),0.75f,1f)));
		production_tile4.setOrigin(0, 0);
//		if(factory.ownShips>MAXSHIPS-1) {
//			production_tile4.setColor(1, 0.1f,0.1f, MathUtils.sin(fade));
//		} else {
			production_tile4.setColor(angle, angle,angle, angle);
//		}
		production_tile4.setRotation(factory.facing.angle());
		production_tile4.setPosition(factory.collisionCenter.x - (35 * factory.facing.x)- (32 * facing90.x), 
				factory.collisionCenter.y	- (35* factory.facing.y) - (32 * facing90.y));
		production_tile4.draw(spriteBatch);
		
		// Draw the needle
		angle = scale_angle(potential_cost);
		if (angle == 0) {
			if (needle_angle > 0 || needle_velocity < 0) {
				needle_velocity = Math.min(needle_velocity + 0.2f * Gdx.graphics.getDeltaTime(), 0.025f);
				needle_angle = Math.max(needle_angle - needle_velocity, 0);
				if (needle_angle == 0) {
					needle_velocity *= Math.pow(-0.475f, Gdx.graphics.getDeltaTime());
				}
			}
		} else {
			needle_velocity = 0;
			needle_angle = angle;
		}
		needle.setOrigin(0, 0);
		needle.setPosition(factory.collisionCenter.x - (2 * factory.facing.x) - (-2 * facing90.x), 
				factory.collisionCenter.y	- (2 * factory.facing.y) - (-2 * facing90.y));
		needle.setRotation(factory.facing.angle() + ((-needle_angle) * 360) - 90);
		needle.draw(spriteBatch);

		production2.setOrigin(0, 0);
		production2.setRotation(factory.facing.angle());
		production2.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
				factory.collisionCenter.y	- (35 * factory.facing.y) - (32 * facing90.y));
		production2.draw(spriteBatch);

		production3.setOrigin(0, 0);
		production3.setRotation(factory.facing.angle());
		production3.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
				factory.collisionCenter.y	- (35 * factory.facing.y) - (32 * facing90.y));
		production3.draw(spriteBatch);
		
		
		// Draw the preview outline
		if (factory.button_held) {
			if (potential_cost > upgradeCost) {
				upgrade_outline.setOrigin(0, 0);
				upgrade_outline.setRotation(factory.facing.angle());
				upgrade_outline.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				upgrade_outline.draw(spriteBatch);
				currentBuildingUnit = 3;
			} else if (potential_cost > frigateCost) {
				frigate_outline.setOrigin(0, 0);
				frigate_outline.setRotation(factory.facing.angle());
				frigate_outline.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32* facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				frigate_outline.draw(spriteBatch);
				currentBuildingUnit = 2;
			} else if (potential_cost > bomberCost) {
				bomber_outline.setOrigin(0, 0);
				bomber_outline.setRotation(factory.facing.angle());
				bomber_outline.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				bomber_outline.draw(spriteBatch);
				currentBuildingUnit = 1;
			} else if (potential_cost > fighterCost) {
				fighter_outline.setOrigin(0, 0);
				fighter_outline.setRotation(factory.facing.angle());
				fighter_outline.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				fighter_outline.draw(spriteBatch);
				currentBuildingUnit = 0;
			} else {
				currentBuildingUnit = -1;
			}
		} else {
			currentBuildingUnit = -1;
			
			float health = factory.healthPercentage();
			if (health < Constants.lowHealthThreshold) {
				float factor = health / Constants.lowHealthThreshold;
				health_none.setOrigin(0, 0);
				health_none.setRotation(factory.facing.angle());
				health_none.setColor(1, factor * 0.3f, factor * 0.3f, 1);
				health_none.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				health_none.draw(spriteBatch);
			} else if (health < Constants.highHealthThreshold) {
				float factor = (health - Constants.lowHealthThreshold) / (Constants.highHealthThreshold - Constants.lowHealthThreshold);
				health_some.setOrigin(0, 0);
				health_some.setRotation(factory.facing.angle());
				health_some.setColor(1, factor * 0.7f + 0.3f, factor * 0.2f + 0.3f, 1);
				health_some.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32*facing90.y));
				health_some.draw(spriteBatch);
			} else {
				float factor = (health - Constants.highHealthThreshold) / (1 - Constants.highHealthThreshold);
				health_full.setOrigin(0, 0);
				health_full.setRotation(factory.facing.angle());
				health_full.setColor((1 - factor) * 0.3f + 0.7f, 1, factor * 0.4f + 0.6f, 1);
				health_full.setPosition(factory.collisionCenter.x - (35 * factory.facing.x) - (32 * facing90.x), 
						factory.collisionCenter.y - (35 * factory.facing.y) - (32 * facing90.y));
				health_full.draw(spriteBatch);
			}
		}

	}
	
	  /**
	   * Normalize a value to exist between 0 and 1 (inclusive).
	   * Mathematically the opposite of lerp(), figures out what proportion
	   * a particular value is relative to start and stop coordinates.
	   */
	  public float norm(float value, float start, float stop) {
	    return (value - start) / (stop - start);
	  }
}

File: demos/pax-britannica/pax-britannica-android/assets/data/scripts/production.lua
local gl = require 'gl'
local blueprints = require 'blueprints'
local v2 = require 'dokidoki.v2'

BUILDING_SPEED = 3

local SEGMENTS = 32
local RADIUS = 32
local DRAW_OFFSET = -4
local SPAWN_OFFSET = 47

UNIT_COSTS = { fighter = 50, bomber = 170, frigate = 360, upgrade = 1080 }

local potential_cost = 0
local needle_angle = 0
local needle_velocity = 0

local texcoord_scroller = -0.25

button_held = false
halt_production = false

local function spawn(unit_type)
  self.resources.amount = self.resources.amount - UNIT_COSTS[unit_type]
  
  if (unit_type == 'upgrade') then 
    self.resources.harvest_rate = self.resources.harvest_rate + 0.25
  else
    local spawn_pos = self.transform.pos + SPAWN_OFFSET * self.transform.facing
    game.log.record_spawn(blueprints[unit_type])
    game.actors.new(blueprints[unit_type],
      {'transform', pos=spawn_pos, facing=self.transform.facing},
      {'ship', player=self.ship.player})
  end
end

function update()  
  if button_held then
    if (potential_cost == 0 and self.resources.amount >= potential_cost) then 
      potential_cost = UNIT_COSTS.fighter 
    end
    if (self.resources.amount > potential_cost + BUILDING_SPEED-1) then
      potential_cost = potential_cost + BUILDING_SPEED
    end
  else
    if potential_cost > UNIT_COSTS.upgrade then
      spawn('upgrade')    
    elseif potential_cost > UNIT_COSTS.frigate then
      spawn('frigate')  
    elseif potential_cost > UNIT_COSTS.bomber then
      spawn('bomber')
    elseif potential_cost > UNIT_COSTS.fighter then
      spawn('fighter')
    end
    potential_cost = 0
  end
  
  -- this is debugging code and should be removed sometime soon...
  local key_z = game.debug_keys.key_pressed(string.byte('Z'))
  local key_x = game.debug_keys.key_pressed(string.byte('X'))
  local key_c = game.debug_keys.key_pressed(string.byte('C'))
  local key_v = game.debug_keys.key_pressed(string.byte('V'))

  if key_z then spawn('fighter') end
  if key_x then spawn('bomber')  end
  if key_c then spawn('frigate') end
  if key_v then spawn('upgrade') end
end

local function scale_angle(frames)
  local angle = 0
  if (frames < UNIT_COSTS.fighter) then
    angle = 0.25
  elseif (frames < UNIT_COSTS.bomber) then
    angle = (frames - UNIT_COSTS.fighter) / (UNIT_COSTS.bomber - UNIT_COSTS.fighter) * 0.25 + 0.25
  elseif (frames < UNIT_COSTS.frigate) then
    angle = (frames - UNIT_COSTS.bomber) / (UNIT_COSTS.frigate - UNIT_COSTS.bomber) * 0.25 + 0.5
  else
    angle = (frames - UNIT_COSTS.frigate) / (UNIT_COSTS.upgrade - UNIT_COSTS.frigate) * 0.25 + 0.75
  end
  return math.min(angle - 0.25, 1)
end

local function get_resources_spent(frames)
  local spent
  if (frames < UNIT_COSTS.fighter) then
    spent = 0
  elseif (frames < UNIT_COSTS.bomber) then
    spent = UNIT_COSTS.fighter
  elseif (frames < UNIT_COSTS.frigate) then
    spent = UNIT_COSTS.bomber
  elseif (frames < UNIT_COSTS.upgrade) then
    spent = UNIT_COSTS.frigate
  else
    spent = UNIT_COSTS.upgrade
  end
  return spent
end

function draw()
  if halt_production then
    return
  end

  local pos = self.transform.pos;
  local offset = DRAW_OFFSET * self.transform.facing

  gl.glPushMatrix()
  gl.glTranslated(pos.x + offset.x, pos.y + offset.y, 0)
  
  game.resources.production_layer_1:draw()
  
  -- Draw the available resources pie-slice
  gl.glBegin(gl.GL_TRIANGLE_FAN)
    local angle = scale_angle(self.resources.amount) * math.pi * 2
    local filled_angle = math.pi * 2 - angle
    gl.glColor4d(0, 0, 0, 0.6)
    gl.glVertex2d(0, 0)
    for point = 0,SEGMENTS do
      local vert = v2.unit(math.pi/2 - angle - point / SEGMENTS * filled_angle) * RADIUS
      gl.glVertex2d(vert.x, vert.y)
    end
    gl.glVertex2d(0, 0)
  gl.glEnd()   
  
  -- Draw the actually used resources slice (if any)
  if self.resources.amount > UNIT_COSTS.fighter and button_held then
    gl.glBegin(gl.GL_TRIANGLE_FAN)
      local cost = get_resources_spent(potential_cost)
      local bottom_highlight_angle = scale_angle(cost) * math.pi * 2
      gl.glColor4d(0.5, 1, 1, 1)
      gl.glVertex2d(0, 0)
      for point = 0,SEGMENTS do
        gl.glVertex2d(math.sin(point / SEGMENTS * math.pi * 0.5 + bottom_highlight_angle) * RADIUS, math.cos(point / SEGMENTS * math.pi * 0.5 + bottom_highlight_angle) * RADIUS)
      end
    gl.glEnd()    
  else
    -- TODO : Grey out the entire thing, or close doors, whatever
  end
  
  -- Draw the needle
  local angle = scale_angle(potential_cost)
  if angle == 0  then
    if needle_angle > 0 or needle_velocity < 0 then
      needle_velocity = math.min(needle_velocity + 0.002, 0.025)
      needle_angle = math.max(needle_angle - needle_velocity, 0)
      if (needle_angle == 0) then needle_velocity = needle_velocity * -0.475 end
    end
  else
    needle_velocity = 0
    needle_angle = angle
  end
  gl.glColor3d(1, 1, 1)
  gl.glPushMatrix()
    gl.glScaled(0.5, 0.5, 1)
    gl.glRotated(-needle_angle * 360, 0, 0, 1)
    game.resources.needle_sprite:draw()
  gl.glPopMatrix()
  
  game.resources.production_layer_2:draw()
  game.resources.production_layer_3:draw()
  
  -- Draw the preview outline 
  if button_held then
    if potential_cost > UNIT_COSTS.upgrade then
      game.resources.upgrade_preview_sprite:draw()
    elseif potential_cost > UNIT_COSTS.frigate then
      game.resources.frigate_preview_sprite:draw() 
    elseif potential_cost > UNIT_COSTS.bomber then
      game.resources.bomber_preview_sprite:draw() 
    elseif potential_cost > UNIT_COSTS.fighter then
      game.resources.fighter_preview_sprite:draw()
    end
  else
    texcoord_scroller = texcoord_scroller + 0.01
    if texcoord_scroller > 0.25 then texcoord_scroller = texcoord_scroller - 0.5 end    
    gl.glMatrixMode(gl.GL_TEXTURE)
    gl.glTranslated(texcoord_scroller, 0,0)
    gl.glMatrixMode(gl.GL_MODELVIEW)
    gl.glPushMatrix()
    gl.glScaled(1.20, 1, 1)
    
    local health = self.ship.health_percentage()  
    if health < game.constants.low_health_threshold then
      local factor = health / game.constants.low_health_threshold
      gl.glScaled(1, math.max(0.5, factor), 1)
      gl.glColor3d(1, factor*0.3, factor*0.3)
      game.resources.health_none:draw()
    elseif health < game.constants.high_health_threshold then
      local factor = (health - game.constants.low_health_threshold) / (game.constants.high_health_threshold - game.constants.low_health_threshold)
      gl.glScaled(1, factor * 0.5 + 0.5, 1)
      gl.glColor3d(1, factor * 0.7 + 0.3, factor * 0.2 + 0.3)
      game.resources.health_some:draw()
    else
      local factor = (health - game.constants.high_health_threshold) / (1 - game.constants.high_health_threshold)
      gl.glScaled(1, health, 1)
      gl.glColor3d((1-factor)*0.3+0.7, 1, factor * 0.4 + 0.6)    
      game.resources.health_full:draw()
    end
    
    gl.glPopMatrix()
    gl.glMatrixMode(gl.GL_TEXTURE)
    gl.glLoadIdentity()
    gl.glMatrixMode(gl.GL_MODELVIEW)      
  end
  
  gl.glColor3d(1, 1, 1)
  
  gl.glPopMatrix() -- main push
end
